### PR TITLE
Refactor Website Structure :  Added Loader, Pagination, Search, and SearchAPI

### LINKED ISSUE #56: Implement reusable pagination component
Like the approach for [a shared component for all items](https://github.com/OWASP/Nest/issues/54), we want a pagination component that works across all pages (Contribute, Projects, Chapters, and Committees).

This pagination component should be flexible and configurable, allowing it to connect with the backend and provide the right pagination features for each page.

### PR DESCRIPTION
**Pull request by Rajgupta36**
Resolves #56  #55 

## Description  
This PR includes the following changes:  
1. Introduced a loader for our website
1. Implemented a reusable pagination component.  
2. Implemented a reusable search component
3. Implemented a reusable Searchapi component
5. Integrated the pagination and searching component into the all pagea.  
6. updated the types 
7. Update test cases

## Checklist  
- [x] Pagination component is reusable.  
- [x] Pagination is functional on the all pages.
- [x] Searchcomponent is reusable.   
- [x] Seachapiendpoint correctly working 
- [x] Loader is working
- [x] no extra api calls 
- [x] updated test cases

## media
----
### pagination component
light mode
image
 ---
 dark mode
  image
  image
---
### SearchBar component
image
---
Loader component
image
image
---
working

https://github.com/user-attachments/assets/544654bc-608e-4258-ad7c-41e23dd36c1d

### CODE REVIEW

**Review by arkid15r — COMMENTED** [review]
Looking forward to reviewing another version of implementation as discussed in Slack 👍

File: frontend/src/components/Search.tsx [hunk 1]
```diff
@@ -1,141 +1,53 @@
...
-import { ProjectDataType, CommitteeDataType } from '../lib/types'
-
-interface SearchBarProps<T> {
-  placeholder: string
-  searchEndpoint: string
+import { debounce } from 'lodash'
+import { Search, X } from 'lucide-react'
+import React, { useCallback, useState } from 'react'
+interface SearchProps {
   // eslint-disable-next-line no-unused-vars
```
Comment (Rajgupta36) [inline frontend/src/components/Search.tsx:5]: cause it only use for state update

File: frontend/src/components/Search.tsx [hunk 2]
```diff
@@ -1,141 +1,53 @@
...
-  onSearchResult: (results: T | null) => void
-  defaultResults: T | null
-  currentPage: number
+  onSearch: (query: string) => void
```
Comment (Rajgupta36) [inline frontend/src/components/Search.tsx:6]: eslint disable cause it is used for state updation only

File: frontend/src/components/Search.tsx [hunk 3]
```diff
@@ -1,129 +1,53 @@
...
-  placeholder: string
-  searchEndpoint: string
+import { debounce } from 'lodash'
+import { Search, X } from 'lucide-react'
+import React, { useCallback, useState } from 'react'
+interface SearchProps {
   // eslint-disable-next-line no-unused-vars
-  onSearchResult: (results: ProjectDataType | null) => void
-  defaultResults: ProjectDataType | null
+  onSearch: (query: string) => void
```
Comment (Rajgupta36) [inline frontend/src/components/Search.tsx:6]: eslint disable cause it is used for state updation only

File: frontend/src/components/Pagination.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
+import { MoreHorizontal } from 'lucide-react'
+import React from 'react'
+
+interface PaginationProps {
+  currentPage: number
+  totalPages: number
+  isLoaded: boolean
+  // eslint-disable-next-line no-unused-vars
+  onPageChange: (page: number) => void
```
Comment (Rajgupta36) [inline frontend/src/components/Pagination.tsx:9]: eslint disable cause it is used for state updation only

File: frontend/src/components/Search.tsx [hunk 5]
```diff
@@ -1,129 +1,53 @@
...
-      } catch (err) {
-        console.error('Search error:', err)
-        setError('Failed to fetch search results. Please try again.')
-        onSearchResult(defaultResults)
-      } finally {
-        setLoading(false)
-      }
-    },
-    [defaultResults, onSearchResult, searchEndpoint]
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (Rajgupta36) [inline frontend/src/components/Search.tsx:12]: we created this component once that' why i used empty array

File: frontend/src/components/Loader.tsx
Comment (arkid15r) [inline frontend/src/components/Loader.tsx:1]: Could you use `LoadingSpinner` or `LoadingIndicator` as a name.

File: frontend/package.json
Comment (arkid15r) [inline frontend/package.json:1]: I don't see package-lock.json changes though

File: frontend/src/components/Loader.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+interface LoaderProps {
+  imageUrl: string
+}
+
+const Loader: React.FC<LoaderProps> = ({ imageUrl }) => {
+  const dark = imageUrl.replace('white', 'black')
+  return (
+    <div
+      className="relative h-16 w-16 animate-spin rounded-full border-4 border-[#98AFC7] dark:border-white"
```
Comment (arkid15r) [inline frontend/src/components/Loader.tsx]: Could you put it to the page center?

File: frontend/src/components/Search.tsx [hunk 7]
```diff
@@ -1,129 +1,53 @@
...
-        return
-      }
-
-      setLoading(true)
-      setError(null)
-      try {
-        const response = await axios.get(searchEndpoint, {
-          params: { q: searchQuery },
-        })
+const SearchComponent: React.FC<SearchProps> = ({ onSearch }) => {
```
Comment (arkid15r) [inline frontend/src/components/Search.tsx]: We need the `placeholder` as it's different for every page

File: frontend/src/components/Search.tsx [hunk 8]
```diff
@@ -1,129 +1,53 @@
...
-        <p className="mt-2 text-sm text-red-600" role="alert">
-          {error}
-        </p>
-      )}
+        {searchQuery && (
+          <button
+            className="absolute right-2 top-1/2 -translate-y-1/2 rounded-full p-1 hover:bg-gray-100 focus:outline-none focus:ring-2 focus:ring-gray-300"
+            onClick={handleClearSearch}
+          >
+            <X className="h-4 w-4 text-gray-400" />
```
Comment (arkid15r) [inline frontend/src/components/Search.tsx:49]: The focus should still be on the search bar after user clicks on "X"

File: frontend/src/lib/api.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+  endpoint: string,
+  query: string,
+  currentPage: number
+): Promise<T> => {
+  const response = await fetch(`${API_URL}/${endpoint}?q=${query}&page=${currentPage}`)
+  if (!response.ok) {
+    throw new Error(`Failed to fetch data: ${response.statusText}`)
+  }
+  const data: T = await response.json()
+  return data
```
Comment (arkid15r) [inline frontend/src/lib/api.ts]:
```suggestion
  return await response.json()
```

File: frontend/src/pages/Chapters.tsx [hunk 10]
```diff
@@ -1,55 +1,97 @@
...
-            const formattedUrls = handleSocialUrls(chapter.idx_related_urls)
+      <SearchBar onSearch={handleSearch} />
+      {!isLoaded ? (
+        <div className="flex min-h-screen w-full flex-col items-center justify-center p-5 text-text">
+          <Loader imageUrl="../public/img/owasp_icon_white_sm.png" />
+        </div>
+      ) : (
+        <div className="flex h-fit w-full flex-col items-center justify-normal gap-4">
+          {totalPages <= 0 && (
+            <div className="text bg:text-white m-4 text-xl"> no Pages found </div>
```
Comment (arkid15r) [inline frontend/src/pages/Chapters.tsx]:
```suggestion
            <div className="text bg:text-white m-4 text-xl">No chapters found</div>
```

File: frontend/src/pages/Chapters.tsx [hunk 11]
```diff
@@ -1,55 +1,97 @@
...
-          chapterData.chapters.map((chapter, index) => {
-            const formattedUrls = handleSocialUrls(chapter.idx_related_urls)
+      <SearchBar onSearch={handleSearch} />
+      {!isLoaded ? (
+        <div className="flex min-h-screen w-full flex-col items-center justify-center p-5 text-text">
+          <Loader imageUrl="../public/img/owasp_icon_white_sm.png" />
+        </div>
+      ) : (
+        <div className="flex h-fit w-full flex-col items-center justify-normal gap-4">
+          {totalPages <= 0 && (
```
Comment (arkid15r) [inline frontend/src/pages/Chapters.tsx]:
```suggestion
          {totalPages === 0 && (
```

File: frontend/src/pages/Chapters.tsx [hunk 12]
```diff
@@ -1,55 +1,97 @@
...
+import { loadData } from '../lib/api'
 import FontAwesomeIconWrapper from '../lib/FontAwesomeIconWrapper'
-import { ChapterDataType } from '../lib/types'
+import { ChapterDataType, ChapterType } from '../lib/types'
 import { handleSocialUrls } from '../lib/utils'
-import { API_URL } from '../utils/credentials'
 
-export default function Chapters() {
-  const [chapterData, setChapterData] = useState<ChapterDataType | null>(null)
+const ChaptersComponent = () => {
```
Comment (arkid15r) [inline frontend/src/pages/Chapters.tsx]:
```suggestion
const ChaptersPage= () => {
```

**Review by arkid15r — CHANGES_REQUESTED** [review]
Great work!
It seems the very first load still 
Screenshot 2024-12-18 at 9 10 41 AM
makes 2 requests

I commented on some components but changes should be made on all of them:

File: frontend/package.json
Comment (Rajgupta36) [reply frontend/package.json:1]: added

File: frontend/src/components/Loader.tsx (hunk 6 above)
Comment (Rajgupta36) [reply frontend/src/components/Loader.tsx]: done

File: frontend/src/components/Search.tsx (hunk 7 above)
Comment (Rajgupta36) [reply frontend/src/components/Search.tsx]: done

File: frontend/src/components/Search.tsx (hunk 8 above)
Comment (Rajgupta36) [reply frontend/src/components/Search.tsx:49]: done

File: frontend/src/lib/api.ts (hunk 9 above)
Comment (Rajgupta36) [reply frontend/src/lib/api.ts]: done

**Comment by Rajgupta36** [comment]
> Great work! It seems the very first load still Screenshot 2024-12-18 at 9 10 41 AM makes 2 requests
> 
> I commented on some components but changes should be made on all of them:

it's cause of react strictmode

File: frontend/src/lib/api.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import { API_URL } from '../utils/credentials'
+export const loadData = async <T>(
+  endpoint: string,
+  query: string,
+  currentPage: number
+): Promise<T> => {
+  const response = await fetch(`${API_URL}/${endpoint}?q=${query}&page=${currentPage}`)
```
Comment (arkid15r) [inline frontend/src/lib/api.ts]:
```js
fetch('https://example.com?' + new URLSearchParams({
    foo: 'value',
    bar: 2,
}).toString())
```

**Review by arkid15r — COMMENTED** [review]
LGTM

Just a small suggestion:

File: frontend/src/lib/api.ts (hunk 13 above)
Comment (Rajgupta36) [reply frontend/src/lib/api.ts:7]: it's already wrapped up in ` ` string

Comment (arkid15r) [reply frontend/src/lib/api.ts]: right, I suggest to create a mapping for better readability. a separate variable would be fine too.

Comment (Rajgupta36) [reply frontend/src/lib/api.ts]:
```
 const response = await fetch(
    `${API_URL}/${endpoint}?` +
      new URLSearchParams({
        q: query,
        page: currentPage.toString(),
      }).toString()
  )
```
updated

**Review by arkid15r — APPROVED** [review]
LGTM

This is major step, now we have search+pagination!

Thank you @Rajgupta36

**Pull request merged**